nixos/modules/virtualisation/make-image.nix

```
@@ -0,0 +1,81 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.virtualisation.image;
+in
+{
+  options = {
+    virtualisation.image = {
```

In response to https://github.com/NixOS/nixpkgs/pull/148186/files#r761543750

```suggestion
    virtualisation.image = {
      file = mkOption {
        readOnly = true;
        type = types.package;
        description = ''
          The generated image derivation, for use by other modules and expressions.
        '';
      }
```

something like that.

Ignore the first line non-change. GitHub is weird when it comes to suggesting additions.

---

I don't actually understand this, what does types.package do there?

---

I guess `package` is not great, because it's not something with `meta.description` and all that good stuff, but in `types` it basically just means "something in the store", which is what the image is. That said, let's change it to `types.path` because we don't really need a stringent check anyway, as only this module will set it anyway.

---

Alright, how is what I just pushed?

---

Looks decent.
Next step is to validate and test it. Seems like the existing `make-disk-image.nix` calls cannot simply be replaced because that could cause conflicts when multiple such modules are used. I initially assumed you were writing this for a completely new use case where that is not an issue. If your goal is to replace those calls, we'll have to use more complicated techniques.
If the goal is to keep it simple for a new use case, perhaps you could take inspiration from `nixos/tests/image-contents.nix` although that's using `makeEc2Test`, which it probably doesn't need (much).
Otherwise, refactoring is necessary, but let's wait for a blog post I'm writing about that topic.